### PR TITLE
Error when variable is circularly referenced in type annotation

### LINKED ISSUE #2980: Type alias circular reference is not detected
According to spec, circular references are prohibited for type aliases and should be reported as errors. But in 1.4 it depends on the order of declarations. For example, no error is reported in the following code:
```ts
type typeAlias = typeof varOfAliasedType;
var varOfAliasedType: typeAlias;
```
Error is reported only if order of statements is reversed.

One more example:
```ts
var varOfAliasedType = func();
function func(): typeAlias { return null; }
type typeAlias = typeof varOfAliasedType;
```
Error is reported only if order of the first two statements is reversed.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #2980.

### CODE REVIEW

**Comment by mhegazy** [comment]
:+1:

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2278,12 +2278,19 @@ module ts {
...
             else if (links.type === resolvingType) {
-                links.type = anyType;
-                if (compilerOptions.noImplicitAny) {
-                    let diagnostic = (<VariableLikeDeclaration>symbol.valueDeclaration).type ?
-                        Diagnostics._0_implicitly_has_type_any_because_it_is_referenced_directly_or_indirectly_in_its_own_type_annotation :
-                        Diagnostics._0_implicitly_has_type_any_because_it_is_does_not_have_a_type_annotation_and_is_referenced_directly_or_indirectly_in_its_own_initializer;
-                    error(symbol.valueDeclaration, diagnostic, symbolToString(symbol));
+                if ((<VariableLikeDeclaration>symbol.valueDeclaration).type) {
+                    // Variable has type annotation that circularly references the variable itself
+                    links.type = unknownType;
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: Why the unknown type here as opposed to the 'any' type?  What's the mental model i should use for when i should use the 'unknown' type?  Thanks!  (also a comment in the code explaining this would be useful).

Comment (mhegazy) [inline src/compiler/checker.ts]: This is always an error, i.e. it is something that does not make sense. the other one, the compiler sets to any as it could not find a better type (because of the self reference). the second case is not an error, it is the expected behavior, it is just that if you have --noImplictAny, the compiler warns about cases when it silently gives things an any type.

**Comment by ahejlsberg** [comment]
Hmm, turns out my fix only fixes it for the first repro in #2980. This one is still an issue:
```typescript
var varOfAliasedType = func();
function func(): typeAlias { return null; }
type typeAlias = typeof varOfAliasedType;
```
No error is reported for the example above (and `varOfAliasedType` is given type `any`). However, if the first two lines are reversed an error *is* reported.

The core issue is that we have two rules that are fundamentally at odds: We say variables that circularly reference themselves in their initializers are given type `any` without an error, whereas type aliases that circularly reference themselves are an error. Depending on which we check first we either get an error or we don't, and then because we cache, we don't see the circularity in the second check.

**Comment by ahejlsberg** [comment]
Latest commits improve circularity detection such that errors are reported on each declaration in a circular chain. This fixes the order dependency issues and also makes it easier for the user to understand the circular causality.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2014,6 +2017,34 @@ module ts {
             }
         }
 
+        // Push an entry on the type resolution stack. If an entry with the given target is not already on the stack,
+        // a new entry with that target and an associated result value of true is pushed on the stack, and the value
+        // true is returned. Otherwise, a circularity has occurred and the result values of the existing entry and
+        // all entries pushed after it are changed to false, and the value false is returned. The target object has
+        // no significance other than to provide a unique identity for a particular type resolution result.
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Nice comment! Very clear!

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -2014,6 +2017,34 @@ module ts {
...
 
+        // Push an entry on the type resolution stack. If an entry with the given target is not already on the stack,
+        // a new entry with that target and an associated result value of true is pushed on the stack, and the value
+        // true is returned. Otherwise, a circularity has occurred and the result values of the existing entry and
+        // all entries pushed after it are changed to false, and the value false is returned. The target object has
+        // no significance other than to provide a unique identity for a particular type resolution result.
+        function pushTypeResolution(target: Object): boolean {
+            for (let i = 0; i < resolutionCount; i++) {
+                if (resolutionTargets[i] === target) {
+                    do {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I think it is a little odd to nest these loops. I would move the inner one outside, and put it after the outer one, because it really logically follows the outer one.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -2014,6 +2017,34 @@ module ts {
...
+                        resolutionResults[i++] = false;
+                    }
+                    while (i < resolutionCount);
+                    return false;
+                }
+            }
+            resolutionTargets[resolutionCount] = target;
+            resolutionResults[resolutionCount] = true;
+            resolutionCount++;
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Is resolutionCount different from resolutionTargets.length and resolutionResults.length?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -2014,6 +2017,34 @@ module ts {
...
+            return true;
+        }
+
+        // Pop an entry from the type resolution stack and return its associated result value. The result value will
+        // be true if no circularities were detected, or false if a circularity was found.
+        function popTypeResolution(): boolean {
+            resolutionTargets[--resolutionCount] = undefined;
+            return resolutionResults[resolutionCount];
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Don't you have to pop the value from resolutionResults?

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -2014,6 +2017,34 @@ module ts {
...
+            resolutionTargets[resolutionCount] = target;
+            resolutionResults[resolutionCount] = true;
+            resolutionCount++;
+            return true;
+        }
+
+        // Pop an entry from the type resolution stack and return its associated result value. The result value will
+        // be true if no circularities were detected, or false if a circularity was found.
+        function popTypeResolution(): boolean {
+            resolutionTargets[--resolutionCount] = undefined;
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I think the following is a little clearer instead of this line:
```ts
resolutionCount--;
resolutionTargets.length--;
```

File: src/compiler/checker.ts (hunk 4 above)
Comment (vladima) [inline src/compiler/checker.ts]: i'd say yes, because during the 'pop' operation we don't resize these arrays but rather replace the value with `undefined` so `length` property does not reflect the actual amount of data that is stored in arrays

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Oh yeah you're right. Why don't we resize the arrays? Is that an optimization?

File: src/compiler/checker.ts (hunk 2 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Actually can you write it in docComment format so that quick info and signature help will display it?

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -2308,19 +2346,13 @@ module ts {
 
         function getTypeOfAccessors(symbol: Symbol): Type {
             let links = getSymbolLinks(symbol);
-            checkAndStoreTypeOfAccessors(symbol, links);
-            return links.type;
-        }
-
-        function checkAndStoreTypeOfAccessors(symbol: Symbol, links?: SymbolLinks) {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Thanks for cleaning that up

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -7342,10 +7372,9 @@ module ts {
                         if (isContextSensitive(node)) {
                             assignContextualParameterTypes(signature, contextualSignature, contextualMapper || identityMapper);
                         }
-                        if (!node.type) {
-                            signature.resolvedReturnType = resolvingType;
+                        if (!node.type && !signature.resolvedReturnType) {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Why doesn't this one have to change to the new model?

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -2014,6 +2017,34 @@ module ts {
             }
         }
 
+        // Push an entry on the type resolution stack. If an entry with the given target is not already on the stack,
+        // a new entry with that target and an associated result value of true is pushed on the stack, and the value
+        // true is returned. Otherwise, a circularity has occurred and the result values of the existing entry and
+        // all entries pushed after it are changed to false, and the value false is returned. The target object has
+        // no significance other than to provide a unique identity for a particular type resolution result.
+        function pushTypeResolution(target: Object): boolean {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I realize the target object only matters for identity of the result. But I'd rather have one type that gets pushed onto the stack. I noticed below, the things that get pushed can be of type Symbol, SymbolLinks or Signature. I would really prefer to converge on one type if possible, just to keep things organized.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Also, it seems very easy to pass in the same object from two different callers by mistake, even when tracking different properties. I wish we could pass a stronger proxy for the property, rather than passing a different owner object in a way that is unrelated to which property we want to track.

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: I'm going to keep it in this format to stay consistent with the rest of the code base.

File: src/compiler/checker.ts (hunk 8 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: It is already guarded by the ContextChecked flag and will never be re-entered.

File: src/compiler/checker.ts (hunk 2 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I don't think the comment formats have to be consistent at the expense of usability, but very well.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Really we should be using doc comments above every function, and every little bit helps.

File: src/compiler/checker.ts (hunk 4 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I think it is better to just pop. It's simpler.

File: src/compiler/checker.ts (hunk 3 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Done.

File: src/compiler/checker.ts (hunk 4 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Switched to push/pop in latest commit.

**Comment by JsonFreeman** [comment]
I just thought of something interesting, an alternate way to fix this issue. I'm not sure if it's simpler, or if it even works, but it's more similar to the old model, where we track cycles by using the `resolvingType` sentinel. The idea is this: When chasing references, we would essentially allow the cycle to go through twice. On the first reentry, we detect the reentry, and give an error or not, depending on whether it's a variable initializer. But we allow the recursion to keep going so that we repeat every part of the cycle at least once. On the second reentry, we stop the recursion, and return back out. I'm wondering if this is a cleaner solution.

**Comment by JsonFreeman** [comment]
This idea would require two sentinel objects, but I think it is simple, and does not rely on us having to be so careful about what object we are passing to the pushTypeResolution method.

**Comment by ahejlsberg** [comment]
You could probably make something like that work but I doubt it would be simpler. Plus it is technically not in the spirit of lazy initialization of an immutable value because the value actually goes through an observable sequence of changes. I think what we have now is the right solution--or certainly right enough.
